tests/50federation/40devicelists.pl

```diff
@@ -201,6 +201,97 @@
    };
 
 
+test "Server correctly resyncs when client query keys and there is no remote cache",
+   requires => [ local_user_fixture(),
+                 $main::INBOUND_SERVER, $main::OUTBOUND_CLIENT,
+                 federation_user_id_fixture(),
+                 room_alias_name_fixture() ],
+
+   check => sub {
+      my ( $user, $inbound_server, $outbound_client, $creator_id, $room_alias_name ) = @_;
+
+      my ( $room_id );
+
+      my $remote_server_name = $inbound_server->server_name;
+      my $datastore          = $inbound_server->datastore;
+
+      my $room_alias = "#$room_alias_name:$remote_server_name";
+
+      # We return two devices, as there was a bug in synapse which correctly
+      # handled returning one device but not two.
+      my $device_id1 = "random_device_id1";
+      my $device_id2 = "random_device_id2";
+
+
+      # We set up a situation where sytest joins a room with a user without
+      # relaying any device keys, and then a client of synapse requests the keys
+      # for that user. This should cause synapse to do a resync and cache those
+      # keys correctly.
+      my $room = $datastore->create_room(
+         creator => $creator_id,
+         alias   => $room_alias,
+      );
+
+      do_request_json_for( $user,
+         method => "POST",
+         uri    => "/r0/join/$room_alias",
+
+         content => {},
+      )->then( sub {
```

Comment:
it's generally less boilerplatey to have the server under test create the room and have sytest join it. Indeed there is a single fixture for it, `federated_rooms_fixture`. Could we use that here?